include/graylog_logger/LoggingBase.hpp

```diff
@@ -9,39 +9,132 @@
 
 #pragma once
 
+#include "graylog_logger/LibConfig.hpp"
 #include "graylog_logger/LogUtil.hpp"
-#include <mutex>
+#include "graylog_logger/ThreadedExecutor.hpp"
+#include <sstream>
 #include <string>
 #include <vector>
+#ifdef WITH_FMT
+#include <fmt/format.h>
+#include <tuple>
+#endif
+#include "graylog_logger/MinimalApply.hpp"
+#include <ciso646>
+#include <future>
+#include <thread>
 
 namespace Log {
 
 class LoggingBase {
 public:
   LoggingBase();
   virtual ~LoggingBase();
-  virtual void log(const Severity Level, const std::string &Message);
+  virtual void log(const Severity Level, const std::string &Message) {
+    log(Level, Message, std::vector<std::pair<std::string, AdditionalField>>());
+  }
   virtual void
   log(const Severity Level, const std::string &Message,
-      const std::vector<std::pair<std::string, AdditionalField>> &ExtraFields);
+      const std::vector<std::pair<std::string, AdditionalField>> &ExtraFields) {
+    auto ThreadId = std::this_thread::get_id();
+    Executor.SendWork([=]() {
+      if (int(Level) > int(MinSeverity)) {
+        return;
+      }
+      LogMessage cMsg(BaseMsg);
+      for (auto &fld : ExtraFields) {
+        cMsg.addField(fld.first, fld.second);
+      }
+      cMsg.Timestamp = std::chrono::system_clock::now();
+      cMsg.MessageString = Message;
+      cMsg.SeverityLevel = Level;
+      std::ostringstream ss;
+      ss << ThreadId;
+      cMsg.ThreadId = ss.str();
+      for (auto &ptr : Handlers) {
+        ptr->addMessage(cMsg);
+      }
+    });
+  }
   virtual void log(const Severity Level, const std::string &Message,
-                   const std::pair<std::string, AdditionalField> &ExtraField);
+                   const std::pair<std::string, AdditionalField> &ExtraField) {
+    log(Level, Message,
+        std::vector<std::pair<std::string, AdditionalField>>{
+            ExtraField,
+        });
+  }
+
+#ifdef WITH_FMT
+  template <typename... Args>
+  void fmt_log(const Severity Level, std::string Format, Args... args) {
+    auto ThreadId = std::this_thread::get_id();
+    auto UsedArguments = std::make_tuple(args...);
+    Executor.SendWork([=]() {
+      if (int(Level) > int(MinSeverity)) {
+        return;
+      }
+      LogMessage cMsg(BaseMsg);
+      cMsg.SeverityLevel = Level;
+      cMsg.Timestamp = std::chrono::system_clock::now();
+      auto format_message = [=, &cMsg](const auto &... args) {
```

Comment:
Looks like you only need Format, so perhaps take that by ref too?
```suggestion
      auto format_message = [&Format, &cMsg](const auto &... args) {
```

Reply:
Right!